onnxruntime/contrib_ops/cpu/tokenizer.cc

```diff
@@ -0,0 +1,472 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "tokenizer.h"
+#include "onnx/defs/schema.h"
+#include "core/common/common.h"
+#include "core/framework/tensor.h"
+
+#include "core/common/utf8_util.h"
+
+#include <codecvt>
+#include <locale>
+
+namespace onnxruntime {
+namespace contrib {
+
+using namespace utf8_util;
+
+ONNX_CPU_OPERATOR_TYPED_MS_KERNEL(
+    Tokenizer,
+    1,
+    string,
+    KernelDefBuilder()
+        .TypeConstraint("T", DataTypeImpl::GetTensorType<std::string>()),
+    contrib::Tokenizer);
+
+namespace tokenizer_details {
+
+const char start_text = 0x2;
+const char end_text = 0x3;
+
+const std::string conv_error("Conversion Error");
+const std::wstring wconv_error(L"Conversion Error");
+
+// Use a Trie like structure for searching multiple strings
+// at once but convert it to a ternary tree for saving space.
+// We insert separators in the same order they are specified.
+// Template parameter is a CharT which can be a char/wchar_t
+// or anything else that supports operator ><,== as long as
+// this is not a variable length sequence. We convert utf8 to utf16
+// before inserting.
+// Value is a supplementary information useful for search hit
+// and is present in the nodes that terminate the whole search pattern
+template <class CharT, class Value>
+class TernarySearchTree {
```

Comment:
Is this something we could re-use from a third-party library or may be we were trying to avoid a OSS registration?

Reply:
OSS is one thing but it was easy to implement and I needed it a little custom at first.

---
In reply to: [238895003](https://github.com/Microsoft/onnxruntime/pull/31#discussion_r238895003) [](ancestors = 238895003)